discovery.md

```diff
@@ -101,18 +101,295 @@ be considered an event subscriber and a target user of this API.
+This API is specified as a REST API with well defined entities and relationships
+between those entities.
 
-### Provider entity Attributes
+All responses will have a content-type of `application/json` and are either
+a single JSON encoded object or an array of objects.
+
+### Entities in the API
+
+1. `Provider`
+2. `Type`
+3. `Producer`
+
+The `Provider` and `Type` entities form the basis of the directory and can be
+used to build user experiences around discovery. The `Producer` entity is
+keyed off of `{Provider.Name, Type.Name}` and provides the necessary details to
+create a subscription for events of that `ce-type` from the selected `provider`.
+
+### REST Paths
+
+Each path in the REST API represents either represents a list (or search) over
+and entity class, or the retrieval of an individual entity by name. Each of
+these operations MUST be supported by compliant discovery providers.
+
+#### `/types?matching=[search term]`
+
+Returns a list of all types in the discovery system, optionally filtering on
+a provided search term (`matching`).
+
+##### matching
+
+* Type: `string`
+* Description: Search term that provides case insensitive match against `type`
+  names. The parameter can match any portion of the type name.
+* Constraints:
+  * OPTIONAL
+  * If present, must be non-empty
+* Examples:
+  * `"com.storage.object"`
+    * matches:
+      * `"com.storage.object.create"`
+      * `"com.storage.object.delete"`
+      * `"com.storage.object.update"`
+
+##### Returns
+
+A JSON array of `type` entities.
+
+#### `/types/{name}`
+
+Retrieves an individual type entity by exact match on the `type` name.
+
+##### Returns
+
+A `type` entity as a JSON object.
+
+#### `/types/{name}/providers`
+
+Retrieves the details about providers that offer the specified `type`.
+
+##### Returns
+
+A JSON array of `provider` entities.
+
+#### `/providers?matching=[search term]`
+
+Returns a list of all providers int he discovery system, optionally filtering on
+a provided search term (`matching`).
+
+##### matching
+
+* Type: `string`
+* Description: Search term that provides case insensitive match against
+  `provider` names. The parameter can match any portion of the provider name.
+* Constraints:
+  * OPTIONAL
+  * If present, must be non-empty
+* Examples:
+  * `"storage"`
+    * matches:
+      * `"Awesome Cloud Storage"`
+      * `"File storage system"`
+      * `"storage Storage STORAGE"`
+
+##### Returns
+
+A JSON array of `provider` entities.
+
+#### `/providers/{name}`
+
+Retrieves an individual provider entity by exact match on the `provider` name.
+
+##### Returns
+
+A `provider` entity as a JSON object.
+
+#### `/providers/{name}/types`
+
+Retrieves the details about the types that are offered by the specified
+`provider`.
+
+##### Returns
+
+A JSON array of `type` entities
+
+#### `/producer/{provider.name}/{type.name}`
+
+Retrieves the `producer` entity that specifies the information necessary to
```

Comment:
can there be more than one Producer for a given Provider and Type?

Reply:
I'd argue it's a valid use case. One example I can think of: exposing an audit trail as a particular event type, which could span many different sources, producers, etc.